parcels/particlesets/baseparticleset.py

```diff
@@ -23,6 +23,13 @@ class NDCluster(ABC):
 # == class BaseParticleSetNDCluster):
 # ==     def __init__(self, collection_type): # this is NOT an abstract method then
 # ==         assert inspect.isclass(collection_type, ParticleCollection)
+# == Response RB: Yes, I think I agree with option (a). This disentangles the inheritance-structure quite a bit. It
+# ==              would mean that collections and particlesets form their own separate hierarchies (so no ParticleSet
+# ==              is a subclass of a collection), bounded together by a "has-a"-relation, meaning that a ParticleSet
+# ==              always 'carries' an instance of a (Particle)Collection. However, this has an impact on the data-access:
+# ==              Currently you call e.g. `p = pset.data_accessor().set_index(3)`, in the old inheritance-idea this would
+# ==              have become something like `p = pset.get_by_index(3)`, but with this suggestion that would change to
+# ==              `p = pset.collection.get_by_index(3)`.
```

Comment:
well, the actual implication is as such: if we do (a), then we need to replicate quite some data-access functions into the collection as forward-methods of particle set into its collection. So there will be some wrapper functions that some (but not me) would consider 'code replication'. I wouldn't be too nutty on that.

Option (b) is a little work-around some people would call a 'hack'. Still this would just be a minute formality, and considering the many 'hacks' that parcels already does for making the C-interface with arbitrary particles work, I'd say (b) is still quite a viable option.

Note to you @RoelBrouwer : it would be good for you to consider that we are probably (i) not gonna get around some implementation tricks to make this work, due to the C-embedding and the strict 'all MPI processes are workers' paradigm for the MPI simplicity (-> see my tricky workaround on ID management in the ID generator for lists) and that (ii) we are also not fixing all the 'hacks' in parcels. Target is: make it work, make it usable, make it adaptive, make it more manageable than now - not: make it UML-MVC conform (cause that is not gonna happen).

I'm okay with (a) as well as (b), but I do have a preference for (b) cause it would auto-enforce the implementation-via-inheritance of the required abstract functions in the specific ParticleSet implementation, where it counts.

Reply:
we shall leave this discussion open as we may return to this later ...